### ISSUE TITLE
Let nativeDotProduct iterate over all query fields if no field is given as argument

### ISSUE THREAD

**Issue by bratseth** [vespa-engine/vespa#6193]

**Comment by havardpe** [comment]
https://github.com/vespa-engine/vespa/pull/6195

**Comment by havardpe** [comment]
available in vespa 6.248.31

**Comment by vandit-thakkar** [comment]
That's a really good feature you have added. I have a question on how internally nativeDotProduct(field) works, basically in one of the experiments, we found this.

We wanted to boost (add score) on some query term matching in some field.

Basically, we had 40 fields which could be used for ranking and based on some model, dynamically per query some tuples of` (<field>,<term>,<weight>)` used to generate (based on raw query) saying, if for a document, `<term>` is found in `<field>`, then add `<weight>` in relevance score for that document. 

So, we wrote 

matches(title,1)*query(title_1)+
matches(title,2)*query(title_2)+
….
matches(title,20)*query(title_20)+

matches(text,1)*query(text_1)+
matches(text,2)*query(text_2)+
matches(text,3)*query(text_3)+
....
matches(text,20)*query(text_20)+

……
……

<same way for rest all fields>
where  `<weight>` is passed as query features.

so we had 800 `matches` feature in ranking expression (assuming we won't have more than 20 terms in the query), so by following 

https://github.com/vespa-engine/vespa/blob/b8e9cbb6cdad643d937de3ffbe700fa8114a8a30/searchlib/src/vespa/searchlib/features/matchesfeature.cpp

It can be concluded that for a document, all 800 features are evaluated, (let's call it experiment 1)

Now, I came to know that `nativeDotProduct(field)` does the same thing, and now tried out with ranking expression.

nativeDotProduct(title)+nativeDotProduct(text)+... <same way for all 40 fields>

now, referring to https://github.com/vespa-engine/vespa/blob/f562d35d3636cfbb6fd6d9caeb94d97dcdfc0f8e/searchlib/src/vespa/searchlib/features/native_dot_product_feature.cpp
[for nativeDotProduct(field)], though in this ranking expression only 40 features are evaluated, each feature takes all 20 query terms in consideration, so in getTermFieldHandle() is apparently called 800 times here also.   (let's call it experiment 2)

To my surprise, I find experiment 2 way faster performant than experiment 1. 

Basically, it looks like somehow same thing happening internally, it's just that `nativeDotProduct(field)` kind of masks `matches(field,1)*query(field_1)+matches(field,2)*query(field_2)+ ... + matches(field,20)*query(field_20)`. Then why I find very huge difference? What optimizations are done for `nativeDotProduct(field)`.

It's to note that I was using double type numbers in experiment 1, while I am restricted to use integer type numbers in experiment 2. But I don't think it could be this much difference maker. (getting 10 times better performance)

**Comment by havardpe** [comment]
Think of the ranking expression as a function that combines features. The function itself is compiled to machine code and is very fast. Each feature used in the function is calculated by separate components called feature executors. The speed increase is due to fewer features being used, resulting in fewer feature executors being run. Also, getTermFieldHandle is called during setup and not per hit, so this should not impact performance when you have a reasonable amount of documents.

**Comment by vandit-thakkar** [comment]
Yes, I do understand that `getTermFieldHandle()` is called only at setup time. So, we can say that each termData has context saved which fields are searched by term for that query, which is found by `<term_data>->lookupField(uint32_t fieldId)`. 

Thanks for the clarification.

**Comment by havardpe** [comment]
Yes, each term can search multiple fields. Match information used for ranking is available separately for each term-field combination. The lookupField(fieldId) is a convenience to avoid having to inspect all fields for a term when you already know which field you are interested in.

**Comment by vandit-thakkar** [comment]
@havardpe  how to call this new rolled out feature` nativeDotProduct`. 
expression : ` nativeDotProduct` is failing

**Comment by bratseth** [comment]
6.248 isn't released yet. Unless you built yourself you don't have it. It will be released on Monday.

**Comment by SarathKumar-k** [comment]
@bratseth , has the new version released with this feature?

**Comment by bratseth** [comment]
Yes
https://github.com/vespa-engine/vespa/releases
